src/overlay/SurveyDataManager.cpp

```diff
@@ -0,0 +1,484 @@
+    return mNonce;
+}
+
+bool
+SurveyDataManager::nonceIsReporting(uint32_t nonce)
+{
+    std::lock_guard<std::recursive_mutex> const guard(mMutex);
+    return phase() == SurveyPhase::REPORTING && mNonce == nonce;
+}
+
+bool
+SurveyDataManager::fillSurveyData(TimeSlicedSurveyRequestMessage const& request,
+                                  TopologyResponseBodyV2& response)
+{
+    ZoneScoped;
+    std::lock_guard<std::recursive_mutex> const guard(mMutex);
+
+    if (phase() == SurveyPhase::REPORTING && mNonce == request.nonce &&
+        mSurveyor == request.request.surveyorPeerID)
+    {
+        if (!mFinalNodeData.has_value())
+        {
+            emitInconsistencyError("getSurveyData");
+            return false;
+        }
+
+        response.nodeData = mFinalNodeData.value();
+        response.inboundPeers = fillTimeSlicedPeerDataList(
+            mFinalInboundPeerData,
+            static_cast<size_t>(request.inboundPeersIndex));
+        response.outboundPeers = fillTimeSlicedPeerDataList(
+            mFinalOutboundPeerData,
+            static_cast<size_t>(request.outboundPeersIndex));
+        return true;
+    }
+    return false;
+}
+
+bool
+SurveyDataManager::surveyIsActive()
+{
+    std::lock_guard<std::recursive_mutex> const guard(mMutex);
+    return phase() != SurveyPhase::INACTIVE;
+}
+
+#ifdef BUILD_TESTS
+void
+SurveyDataManager::setPhaseMaxDurationsForTesting(
+    std::chrono::minutes maxPhaseDuration)
+{
+    std::lock_guard<std::recursive_mutex> const guard(mMutex);
+    mMaxPhaseDurationForTesting = maxPhaseDuration;
+}
+#endif
+
+SurveyPhase
+SurveyDataManager::phase()
+{
+    if (mRecordStartTime.has_value() && !mRecordEndTime.has_value())
+    {
+        if (mClock.now() >
+            mRecordStartTime.value() + getRecordingPhaseMaxDuration())
+        {
+            CLOG_INFO(
+                Overlay,
+                "Survey recording phase has expired. Resetting survey data.");
+            reset();
+            return SurveyPhase::INACTIVE;
+        }
+        return SurveyPhase::RECORDING;
+    }
+    if (mRecordStartTime.has_value() && mRecordEndTime.has_value())
+    {
+        if (mClock.now() >
+            mRecordEndTime.value() + getReportingPhaseMaxDuration())
+        {
+            CLOG_INFO(
+                Overlay,
+                "Survey reporting phase has expired. Resetting survey data.");
+            reset();
+            return SurveyPhase::INACTIVE;
+        }
+        return SurveyPhase::REPORTING;
+    }
+    return SurveyPhase::INACTIVE;
+}
+
+void
+SurveyDataManager::reset()
+{
+    mRecordStartTime.reset();
+    mRecordEndTime.reset();
+    mNonce.reset();
+    mSurveyor.reset();
+    mRecordingNodeData.reset();
+    mRecordingInboundPeerData.clear();
+    mRecordingOutboundPeerData.clear();
+    mFinalNodeData.reset();
+    mFinalInboundPeerData.clear();
+    mFinalOutboundPeerData.clear();
+}
+
+void
+SurveyDataManager::emitInconsistencyError(std::string const& where)
```

Comment:
I like the idea of gracefully resetting survey if any of the invariants are violated. I'd also recommend to throw when testing, so that we don't accidentally miss a failure.

Reply:
Done in 0c591103e26d506b3256371758175406874ed123